### PR TITLE
[DSCP-130] Use non-static IV for symmetric encryption scheme

### PR DESCRIPTION
**Pull request by flyingleafe**
Using fixed initial vectors for authenticated encryption schemes is unsafe (see https://en.wikipedia.org/wiki/Initialization_vector).
This PR introduces random generation of IV on each encryption. Yes, this means that encrypting the same plaintext with the same password will yield different ciphertexts, but that's actually the point.
Also, now encryption key is derived from a passphrase using PBKDF2 and not simply a hash, which is more secure.

### CODE REVIEW

File: src/Dscp/Crypto/Encrypt.hs [hunk 1]
```diff
@@ -112,6 +112,10 @@ type CipherType = AES256
 authHeader :: ByteString
 authHeader = ""
 
+-- | Salt used for key derivation from password.
+passGenSalt :: ByteString
+passGenSalt = "dscp-educator-pass-gen"
```
Comment (Martoon-00) [inline src/Dscp/Crypto/Encrypt.hs]: If you use fixed salt then adversary can still generate rainbow table and guess your password.
I may be wrong though. If I understand correctly, we used password hash as salt in similar code.

Comment (flyingleafe) [reply src/Dscp/Crypto/Encrypt.hs]: No, in Cardano we've also used fixed salt for symmetric encryption key derivation (see `Pos.Crypto.HD`).
Generating a rainbow table for given key derivation function is kind of tiresome. The purpose of the _salt_ is to make your own key derivation function unique and different from other ones, thus preventing a rainbow table for your function from being _already available_.

File: src/Dscp/Crypto/Encrypt.hs [hunk 2]
```diff
@@ -125,37 +131,107 @@ authHeader = ""
...
+            keyFromPassPhrase pp
         cipher :: CipherType =
             onCryptoFailure impossible identity $
             cipherInit ppHashKey
     in onCryptoFailure impossible identity $
-       aeadInit aeadMode cipher initIV
+       aeadInit aeadMode cipher iv
 
 -- | Encrypt given 'ByteArray' with AES.
+-- TODO: use Secure Random from OpenSSL for IV calculation instead.
```
Comment (volhovm) [inline src/Dscp/Crypto/Encrypt.hs]: Why do you need it?

File: src/Dscp/Util.hs [hunk 3]
```diff
@@ -32,3 +42,23 @@ wrapRethrowIO
...
+
+toBase :: ByteArrayAccess ba => Base -> ba -> Text
+toBase base = decodeUtf8 @Text @ByteString . convertToBase base
+
+fromBase :: forall ba. FromByteArray ba => Base -> Text -> Either String ba
+fromBase base =
+    convertFromBase base . encodeUtf8 @Text @ByteString >=>
+    fromByteArray @ba @ByteString
+
+toBase64, toHex :: ByteArrayAccess ba => ba -> Text
```
Comment (volhovm) [inline src/Dscp/Util.hs]: We also have it in `serokell-util`. Not saying you necessarily should use it, but...

File: tests/Test/Dscp/Crypto/Serialise.hs [hunk 4]
```diff
@@ -0,0 +1,12 @@
+module Test.Dscp.Crypto.Serialise where
```
Comment (volhovm) [inline tests/Test/Dscp/Crypto/Serialise.hs]: Please also add positive/negative tests for the encryption itself (not for the binary instances only).

**Review by volhovm — CHANGES_REQUESTED** [review]
Also, can you please specify that `Dscp.Crypto.Encrypt` is a symmetric encryption module (in the module's haddock)? In general the PR is ok.

File: tests/Test/Dscp/Crypto/Serialise.hs (hunk 4 above)
Comment (flyingleafe) [reply tests/Test/Dscp/Crypto/Serialise.hs]: They are already added in previous PR, look at the `Test.Dscp.Crypto.Encrypt` module.

Comment (volhovm) [reply tests/Test/Dscp/Crypto/Serialise.hs]: Aha! Alright then.

File: src/Dscp/Util.hs (hunk 3 above)
Comment (flyingleafe) [reply src/Dscp/Util.hs]: Fair enough, but these ones work for every `ByteArrayAccess`, so I'll stick with them.

File: src/Dscp/Crypto/Encrypt.hs (hunk 2 above)
Comment (flyingleafe) [reply src/Dscp/Crypto/Encrypt.hs]: For the same reason why we might need it at all - to make random more _random_ where _randomness_ matter. Do you claim that there's no need for this and using Secure Random(tm) from OpenSSL is simply irrelevant?

**Review by volhovm — APPROVED** [review]

File: src/Dscp/Crypto/Encrypt.hs (hunk 2 above)
Comment (volhovm) [reply src/Dscp/Crypto/Encrypt.hs]: I doubt that we actually need to use openssl. It's not like it's the only way to generate random data properly.

**Pull request merged**